docs: add platform engineering projects 2022 (#275)
bytebase/star-history: public/blog/star-history-yearly-pick-2022-platform-engineering.md

![sourcegraph-search](/blog/assets/yearly-pick-platform-engineering-2022/sourcegraph-search.webp)

[Sourcegraph](https://github.com/sourcegraph/sourcegraph) is a code search platform. Imagine a developer fixing a bug, but needs to jump to another part of the code, so he has to open another file in his IDE, which breaks his current working state. Sourcegraph's web-based code search interface allows developers to keep their current state while exploring other parts of the code, which greatly reduces the cost of switching. Advanced features such as code navigation, code insight help developers onboard new projects faster, understand their projects better (how code changes over time and its current state), as well as the ability to apply code changes in different repositories like Batch Changes.

In addition, the [Sourcegraph Podcast](https://about.sourcegraph.com/podcast), featuring co-founder/CTO Beyang and authors/engineers/DevRel folks of other startups, is an insightful podcast about projects' origin stories, these folks' worldviews, experiences, favourite tools and techs, and more.

## Temporal

![temporal](/blog/assets/yearly-pick-platform-engineering-2022/temporal.webp)

> [Temporal](https://github.com/temporalio/temporal) is another rising stat and serves as a core piece in platform engineering. It is a microservice task scheduling execution platform that executes application logic units, workflows in a resilient way, automatically handles intermittent failures, and retries failed operations. Although the first release only came in 2020.2, the founders built Cadence workflow (uber/cadence) back in Uber to solve Uber's workflow/orchestration problems, and Temporal is a commercialized fork version of Cadence.

Reviewer: ```suggestion
[Temporal](https://github.com/temporalio/temporal) is another rising star and serves as a core piece in platform engineering. It is a microservice task scheduling execution platform that executes application logic units, workflows in a resilient way, automatically handles intermittent failures, and retries failed operations. Although the first release only came in 2020.2, the founders built Cadence workflow (uber/cadence) back in Uber to solve Uber's workflow/orchestration problems, and Temporal is a commercialized fork version of Cadence.
```

Sidenote: Temporal annual conference Relay merch looks so damn good 😍.

![temporal-swag](/blog/assets/yearly-pick-platform-engineering-2022/temporal-swag.webp)

## Terraform

![terraform](/blog/assets/yearly-pick-platform-engineering-2022/terraform.webp)

[Terraform](https://github.com/hashicorp/terraform) is a tool for managing infrastructure (infrastructure-as-code) under HashiCorp. Terraform v0.1 was released back in 2014, but only became the industry leader 8 years later, revolutionizing the entire workflow, and was almost shut down in the early days because of the slow adoption.

Terraform Provider started out as a way to smooth out the differences in managing the services of several large public cloud vendors, and has now become the de facto standard for managing all cloud/SaaS resources.

> ## Platform engineering is the future

Reviewer: ```suggestion
## Platform Engineering is the Future
```

Well, this is it for platform engineering projects, from big names to the starlets, let's review them again:

- The underlying runtime platform - Kubernetes
- The service catalog and developer portal - Backstage
- API gateway - Kong
- Universal Search - Elasticsearch

> - Code hosting and CI/CD - GitLab

Reviewer: ```suggestion
- Code hosting, CI/CD, and DevSecOps - GitLab
```

- Code Search - Sourcegraph
- Monitoring/Alerting - Prometheus
- Dashboard - Grafana
- Task Scheduling and Execution - Temporal

> - Managing all of the above resources (and the others) in code - Terraform

Reviewer: ```suggestion
- Manage all cloud resources and many of the above resources (and the others) in code - Terraform
```

The above projects will constitute a pretty solid platform engineering backbone, while there are a couple of other core components worth mentioning.

> - Security platform (login, VPN, Audit Log, intrusion detection) - [Snyk](https://snyk.io/), [tailscale](https://tailscale.com/).
> - Resource platform (auto-scaling, scheduling, FinOps) [OpenCost](http://opencost.io/)
> - Configuration platform (application configurations, feature flags) [Apollo](https://github.com/apolloconfig/apollo), [Flagsmith](https://flagsmith.com/).
> - Database development platform (database change, query, admin, security management) - [Bytebase](https://bytebase.com)

Reviewer: ```suggestion
- Security and Access Control (network access, software supply chain, intrusion detection, Audit Log) - [Snyk](https://snyk.io/), [tailscale](https://tailscale.com/), [Vault](https://www.hashicorp.com/products/vault), [Boundary](https://www.boundaryproject.io/).
- Resource Allocation (auto-scaling, scheduling, FinOps) - [OpenCost](http://opencost.io/).
- Configuration (application configurations, feature flags) - [Apollo](https://github.com/apolloconfig/apollo), [Flagsmith](https://flagsmith.com/).
- Database Development (database change, query, admin, security) - [Bytebase](https://bytebase.com).
```

## Terraform

![terraform](/blog/assets/yearly-pick-platform-engineering-2022/terraform.webp)

[Terraform](https://github.com/hashicorp/terraform) is a tool for managing infrastructure (infrastructure-as-code) under HashiCorp. Terraform v0.1 was released back in 2014, but only became the industry leader 8 years later, revolutionizing the entire workflow, and was almost shut down in the early days because of the general response.

Terraform Provider started out as a way to smooth out the differences in managing the services of several large public cloud vendors, and has now become the de facto standard for managing various resources.

## Platform engineering is the future

Well, this is it for platform engineering projects, from big names to the newcomers, it feels like this is a tough category, let's take another look:

- The low-level runtime platform - Kubernetes
- The top-level catalog and developer portal - Backstage
- API gateway - Kong
- Universal Search - Elasticsearch
- Code hosting - GitLab
- Code Search - Sourcegraph
- Monitoring/Alerting - Prometheus
- Dashboard - Grafana
- Task Scheduling Execution - Temporal
- Keeping all of these resources in check - Terraform

> Now you have a fully private and deployable platform engineering system, but is there any other core components that have not been mentioned?

Reviewer: ```suggestion
The above projects will constitute a pretty solid platform engineering backbone, while there are a couple of other core components worth mentioning.
```

> - Security platform (login, VPN, Audit Log, intrusion detection)

Reviewer: ```suggestion
- Security platform (login, VPN, Audit Log, intrusion detection) - [Snyk](https://snyk.io/), [tailscale](https://tailscale.com/).
```

> - Resource platform (expansion and contraction, scheduling, FinOps)

Reviewer: ```suggestion
- Resource platform (auto-scaling, scheduling, FinOps) [OpenCost](http://opencost.io/)
```

> - Configuration platform (application configurations, feature flags)

Reviewer: ```suggestion
- Configuration platform (application configurations, feature flags) [Apollo](https://github.com/apolloconfig/apollo), [Flagsmith](https://flagsmith.com/).
```

> - Database development platform (for database change & query management)

Reviewer: ```suggestion
- Database development platform (database change, query, admin, security management) - [Bytebase](https://bytebase.com)
```

> Sidenote: if you are interested in database development platforms, we’d recommend [Bytebase](https://www.bytebase.com/). As for the other items, DM us [@StarHistoryHQ](https://twitter.com/StarHistoryHQ) and we can provide recommendations ;)

Reviewer: ```suggestion
```
